config/unicorn.conf.rb

```diff
@@ -144,25 +144,35 @@ def check_sidekiq_heartbeat
           @sidekiq_next_heartbeat_check ||= Time.new.to_i + @sidekiq_heartbeat_interval
 
           if @sidekiq_next_heartbeat_check < Time.new.to_i
-
-            last_heartbeat = Jobs::RunHeartbeat.last_heartbeat
-            restart = false
+            @sidekiq_next_heartbeat_check = Time.new.to_i + @sidekiq_heartbeat_interval
+            restarted = false
 
             if out_of_memory?
               Rails.logger.warn("Sidekiq is consuming too much memory (using: %0.2fM) for '%s', restarting" % [(max_rss.to_f / 1.megabyte), ENV["DISCOURSE_HOSTNAME"]])
-              restart = true
+              Demon::Sidekiq.restart
+              restarted = true
             end
 
-            if last_heartbeat < Time.new.to_i - @sidekiq_heartbeat_interval
-              STDERR.puts "Sidekiq heartbeat test failed, restarting"
-              Rails.logger.warn "Sidekiq heartbeat test failed, restarting"
-
-              restart = true
+            if !restarted
+              Demon::Sidekiq.heartbeat_queues.each do |queue|
+                last_heartbeat = Demon::Sidekiq.get_queue_last_heartbeat(queue)
+
+                if last_heartbeat < Time.new.to_i - @sidekiq_heartbeat_interval
+                  pid = queue.split("_").last.to_i
+                  STDERR.puts "Sidekiq heartbeat test for worker #{pid} failed, restarting"
+                  Rails.logger.warn "Sidekiq heartbeat test for worker #{pid} failed, restarting"
+                  Demon::Sidekiq.demons.values.each do |demon|
```

Comment:
Since you're only looking for one item in the list, you could do something like this

```ruby
if demon = Demon::Sidekiq.demons.values.find { |d| d.pid == pid }
  demon.stop
  demon.start
  restarted = true
end
```